File: producer4/producer.go
package main

import (
	"fmt"
	"log"
	"os"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

func main() {
	host := getenv("RABBIT_HOST", "rabbitmq")
	user := getenv("RABBIT_USER", "user")
	pass := getenv("RABBIT_PASS", "password")
	url := fmt.Sprintf("amqp://%s:%s@%s:5672/", user, pass, host)

	conn, err := amqp.Dial(url)
	if err != nil {
		log.Fatalf("No se pudo conectar: %v", err)
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		log.Fatalf("Error al crear canal: %v", err)
	}
	defer ch.Close()

	exchange := "demo_exchange"
	ch.ExchangeDeclare(exchange, "fanout", true, false, false, false, nil)

	i := 1
	for {
		msg := fmt.Sprintf("Mensaje Go #%d", i)
		err = ch.Publish(exchange, "", false, false, amqp.Publishing{
			ContentType: "text/plain",
			Body:        []byte(msg),
		})
		if err != nil {
			log.Printf("Error al enviar: %v", err)
		} else {
			log.Printf("[x] Enviado: %s", msg)
		}
		i++
		time.Sleep(3 * time.Second)
	}
}

func getenv(key, fallback string) string {
	val := os.Getenv(key)
	if val == "" {
		return fallback
	}
	return val
}

File: producer3/producer.js
import amqplib from "amqplib";

const RABBIT_HOST = process.env.RABBIT_HOST || "rabbitmq";
const RABBIT_USER = process.env.RABBIT_USER || "user";
const RABBIT_PASS = process.env.RABBIT_PASS || "password";
const EXCHANGE = "demo_exchange";

async function main() {
  const conn = await amqplib.connect(`amqp://${RABBIT_USER}:${RABBIT_PASS}@${RABBIT_HOST}`);
  const ch = await conn.createChannel();
  await ch.assertExchange(EXCHANGE, "fanout", { durable: true });

  let count = 1;
  while (true) {
    const msg = `Mensaje NodeJS #${count}`;
    ch.publish(EXCHANGE, "", Buffer.from(msg));
    console.log(`[x] Enviado: ${msg}`);
    count++;
    await new Promise(r => setTimeout(r, 3000));
  }
}

main().catch(err => console.error("Error:", err));
